lib/cloud/gcp/kubernetes.go

```diff
@@ -0,0 +1,270 @@
+// for Teleport Discovery to decide whether or not to import the cluster.
+type GKECluster struct {
+	// Name is the cluster name.
+	Name string
+	// Description is the cluster description field in GCP.
+	Description string
+	// Location is the cluster location.
+	Location string
+	// ProjectID is the GCP project ID to which the cluster belongs.
+	ProjectID string
+	// Status is the cluster current status.
+	Status containerpb.Cluster_Status
+	// Labels are the cluster labels in GCP.
+	Labels map[string]string
+}
+
+// ClusterDetails is the cluster identification properties.
+type ClusterDetails struct {
+	// ProjectID is the GCP project ID to which the cluster belongs.
+	ProjectID string
+	// Locations are the cluster locations.
+	Location string
+	// Name is the cluster name.
+	Name string
+}
+
+// CheckAndSetDefaults check and set defaults for ClusterDetails.
+func (c *ClusterDetails) CheckAndSetDefaults() error {
+	if len(c.ProjectID) == 0 {
+		return trace.BadParameter("ProjectID must be set")
+	}
+	if len(c.Location) == 0 {
+		return trace.BadParameter("Location must be set")
+	}
+	if c.Location == types.Wildcard {
+		return trace.BadParameter("Location does not support wildcards")
+	}
+	if len(c.Name) == 0 {
+		return trace.BadParameter("Name must be set")
+	}
+	return nil
+}
+
+// toGCPEndpointName generates a GCP endpoint identifier with the following
+// format: projects/*/locations/*/clusters/*.
+func (c *ClusterDetails) toGCPEndpointName() string {
+	return fmt.Sprintf("projects/%s/locations/%s/clusters/%s", c.ProjectID, c.Location, c.Name)
+}
+
+// NewGKEClient returns a GKEClient interface wrapping container.ClusterManagerClient and
+// oauth2.TokenSource for interacting with GCP Kubernetes Service.
+func NewGKEClient(ctx context.Context) (GKEClient, error) {
+	var cfg GKEClientConfig
+	client, err := NewGKEClientWithConfig(ctx, cfg)
+	return client, trace.Wrap(err)
+}
+
+// NewGKEClientWithConfig returns a GKEClient interface wrapping
+// container.ClusterManagerClient and oauth2.TokenSource for interacting with GCP
+// Kubernetes Service.
+func NewGKEClientWithConfig(ctx context.Context, cfg GKEClientConfig) (GKEClient, error) {
+	if err := cfg.CheckAndSetDefaults(ctx); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &gkeClient{cfg}, nil
+}
+
+// gkeClient implements the GKEClient interface by wrapping container.ClusterManagerClient
+// and oauth2.TokenSource for interacting with GCP Kubernetes Service.
+type gkeClient struct {
+	GKEClientConfig
+}
+
+// ListClusters lists the GCP GKE clusters that belong to the projectID and are
+// located in location.
+// location supports wildcard "*".
+func (g *gkeClient) ListClusters(ctx context.Context, projectID string, location string) ([]GKECluster, error) {
+	if len(projectID) == 0 {
+		return nil, trace.BadParameter("projectID must be set")
+	}
+	if len(location) == 0 {
+		return nil, trace.BadParameter("zone must be set")
+	}
+
+	location = convertLocationToGCP(location)
```

Comment:
Nit: This variable is only used in one place so you probably don't need to define it separately and can just use `convertLocationToGCP(location)` inline below.